feat(route): add xinhua-daily-telegraph

Dustin-Jiang opened the pull request:
Add xinhua-daily telegraph route and doc mentioned in issue #11390.

<!-- 
如有疑问，请参考 https://github.com/DIYgod/RSSHub/discussions/8002
Reference: https://github.com/DIYgod/RSSHub/discussions/8002
-->

## 该 PR 相关 Issue / Involved issue

Close #11390

## 完整路由地址 / Example for the proposed route(s)

<!--
请在 `routes` 区域填写以 / 开头的完整路由地址，否则你的 PR 将会被无条件关闭。
如果路由包含在文档中列出可以完全穷举的参数（例如分类），请依次全部列出。

Please include route starts with /, with all required and optional parameters. Fail to comply will result in your pull request being closed automatically.
```route
/some/route
/some/other/route
/dont/use/this/or/modify/it
/use/the/fenced/code/block/below
```
如果你的 PR 与路由无关, 请在 `routes` 区域 填写 `NOROUTE`，而不是直接删除 `routes` 区域。否则你的 PR 将会被无条件关闭。
If your changes are not related to route, please fill in `routes` with `NOROUTE`. Fail to comply will result in your PR being closed.
-->

```routes
/mrdx/today
```

## 新 RSS 检查列表 / New RSS Script Checklist
  
- [x] 新的路由 New Route
  - [x] 跟随 [v2 路由规范](https://docs.rsshub.app/joinus/script-standard.html) Follows [v2 Script Standard](https://docs.rsshub.app/en/joinus/script-standard.html)
- [x] 文档说明 Documentation
  - [x] 中文文档 CN
  - [x] 英文文档 EN
- [x] 全文获取 fulltext
  - [x] 使用缓存 Use Cache
- [ ] 反爬/频率限制 anti-bot or rate limit?
  - [ ] 如果有, 是否有对应的措施? If yes, do your code reflect this sign?
- [x] [日期和时间](https://docs.rsshub.app/joinus/pub-date.html) [date and time](https://docs.rsshub.app/en/joinus/pub-date.html)
  - [x] 可以解析 Parsed
  - [x] 时区调整 Correct TimeZone
- [ ] 添加了新的包 New package added
- [ ] `Puppeteer`

## 说明 / Note

生成《新华每日电讯》报纸电子版文章全文和文内图片。

Dustin-Jiang committed:
feat(route): add xinhua-daily-telegraph

Add xinhua-daily telegraph route and doc mentioned in issue #11390.

Dustin-Jiang committed:
feat(route): add radar for xinhua-daily-telegraph

Add required radar config file.

TonyRL commented on lib/v2/xinhua-daily-telegraph/daily.js:
```diff
@@ -0,0 +1,71 @@
+const { parseDate, getElementChildrenInnerText, getHTMLWithCache } = require('./utils');
+
+module.exports = async (ctx) => {
+    const date = timezone(new Date(), 8);
+    const dateFormatted = date.getFullYear().toString() + ('0' + (date.getMonth() + 1).toString()).slice(-2) + date.getDate().toString();
```
You can use `dayjs` instead of manipulate the date string manually
```js
const dayjs = require('dayjs');
const utc = require('dayjs/plugin/utc');
dayjs.extend(utc);
const dateFormatted = dayjs().utcOffset(8).format('YYYYMMDD'));
```
---
or you can simply visit `http://mrdx.cn/content/PaperIndex.htm` since it will redirect you to the right place

TonyRL commented on lib/v2/xinhua-daily-telegraph/daily.js:
```diff
@@ -0,0 +1,71 @@
+    const baseUrl = `http://mrdx.cn/content/${dateFormatted}`;
+    const link = `${baseUrl}/Page01DK.htm`;
+
+    const $ = cheerio.load(await getHTMLWithCache(link, ctx));
```
Do not cache this page or the feed won't receive any updates since the cache may keep extending its expiry date.
```suggestion
    const response = await got(link);
    const $ = cheerio.load(response.data);
```

TonyRL commented on lib/v2/xinhua-daily-telegraph/daily.js:
```diff
@@ -0,0 +1,71 @@
+
+    const $ = cheerio.load(await getHTMLWithCache(link, ctx));
+
+    const passagesList = $('.listdaohang > ul > li a');
+    const passageLinks = [];
+
+    // 当日所有文章Url在<a>的`daoxiang` attribute中
+    // 其站点使用**非标准**的HTML语法
+    // eg: `<a class="foo" daoxiang="foo-bar.htm">Lorem ipsum</a>`
+    for (let i = 0; i < passagesList.length; i++) {
+        passageLinks.push(passagesList[i].attribs.daoxiang);
+    }
```
```suggestion
    const passagesList = $('.listdaohang > ul > li a')
        .toArray()
        .map((item) => {
            item = $(item);
            return {
                title: item.text(),
                link: new URL(item.attr('daoxiang'), response.url).href,
            };
        });
```

TonyRL commented on lib/v2/xinhua-daily-telegraph/daily.js:
```diff
@@ -0,0 +1,71 @@
+    const results = await Promise.all(
+        passageLinks.map(async (passage) => {
+            const link = `${baseUrl}/${passage}`;
+            const response = await getHTMLWithCache(link, ctx);
```
Caching the HTML response and letting RSSHub process it again every time someone fetches the feed may not be a good idea. Instead, we can cache the processed result and return it.

TonyRL commented on lib/v2/xinhua-daily-telegraph/daily.js:
```diff
@@ -0,0 +1,71 @@
+
+    for (const $ of results) {
+        // 标题在`h2`,`h3`,`h4`,`h5`中，有多余空格和换行符
+        const title =
```
`title` will be trimmed by middleware https://github.com/DIYgod/RSSHub/blob/df521a6f2c4ba91fff781b36e8fe785bb11e7d87/lib/middleware/template.js#L52

TonyRL commented on lib/v2/xinhua-daily-telegraph/daily.js:
```diff
@@ -0,0 +1,71 @@
+            title,
+            author,
+            pubDate: timezone(parseDate(pubDate), 8),
+            guid: link,
```
`guid` must be unique. `link` (defined in line 10) is not unique.

TonyRL commented on lib/v2/xinhua-daily-telegraph/daily.js:
```diff
@@ -0,0 +1,71 @@
+        passageLinks.push(passagesList[i].attribs.daoxiang);
+    }
+
+    const passages = [];
+    const results = await Promise.all(
+        passageLinks.map(async (passage) => {
+            const link = `${baseUrl}/${passage}`;
+            const response = await getHTMLWithCache(link, ctx);
+            return cheerio.load(response);
+        })
+    );
+
+    for (const $ of results) {
+        // 标题在`h2`,`h3`,`h4`,`h5`中，有多余空格和换行符
+        const title =
+            getElementChildrenInnerText($('.bggray > .margin15 > h2')[0]).trim() +
+            getElementChildrenInnerText($('.bggray > .margin15 > h3')[0]).trim() +
+            getElementChildrenInnerText($('.bggray > .margin15 > h4')[0]).trim() +
+            getElementChildrenInnerText($('.bggray > .margin15 > h5')[0]).trim();
+
+        // 作者和发布时间
+        // 中间用超多空格分隔
+        // 头尾均有多余空格
+        // eg: 新华每日电讯     2022年12月21日
+        const [author, pubDate] = getElementChildrenInnerText($('.shijian')[0])
+            .split(' ')
+            .filter((item) => item !== '')
+            .map((item) => item.trim());
+
+        const content = $('.contenttext').html();
+
+        passages.push({
+            title,
+            author,
+            pubDate: timezone(parseDate(pubDate), 8),
+            guid: link,
+            link,
+            description: content,
+        });
+    }
```
```suggestion
    const passages = await Promise.all(
        passagesList.map((passage) =>
            ctx.cache.tryGet(passage.link, async () => {
                const { data: response } = await got(passage.link);
                const $ = cheerio.load(response);

                passage.title = $('.margin15').first().children().not('.shijian').text();

                const match = $('.shijian')
                    .text()
                    .match(/\s+(.+)\n+\s+.*(\d{4}年\d{2}月\d{2}日)/);
                passage.author = match[1];
                passage.pubDate = timezone(parseDate(match[2], 'YYYY-MM-DD'), 8);

                passage.description = $('.contenttext').html();

                return passage;
            })
        )
    );
```

TonyRL commented on lib/v2/xinhua-daily-telegraph/daily.js:
```diff
@@ -0,0 +1,71 @@
+
+    ctx.state.data = {
+        title: '新华每日电讯',
+        link: 'http://mrdx.cn/',
```
```suggestion
        link,
```

TonyRL commented on lib/v2/xinhua-daily-telegraph/daily.js:
```diff
@@ -0,0 +1,71 @@
+        description:
+            '中国报纸发行前三强。位列《人民日报》，《参考消息》之后。《新华每日电讯》是新华社出版的一份新闻电讯报，1993年创办，具有很高的权威性和准确性，有“一报在手，便知天下”之美誉。《新华每日电讯》为对开八版日报，集中刊登[新华社]每天向国内播发的电讯稿及图片稿。打开《新华每日电讯》，中国和全世界每天发生的重大事件便将一目了然。在人类生活节奏日益加快的当今世界，《新华每日电讯》以最便捷、最醒目的方式为公众提供最重要的新闻报道，受到公众的喜爱和拥护。',
+    };
+
+    return;
```
```suggestion
```

TonyRL reviewed (changes requested):
Please use path `/mrdx/today` instead

Dustin-Jiang committed:
feat(router): change router name and cache method

Change router name from `xinhua-daily-telegraph` to `mrdx`.
Fix cache method to avoid processing at each request.

TonyRL commented on lib/v2/mrdx/daily.js:
```diff
@@ -0,0 +1,80 @@
+    const responses = await Promise.all(
+        passagesList.map(async (passage) => {
+            // Return and cache responses processed by cheerio.
+            const response = await getHTMLWithCache(passage.link, ctx);
```
> Caching the HTML response and letting RSSHub process it again every time someone fetches the feed may not be a good idea. Instead, we can cache the processed result and return it.

https://github.com/DIYgod/RSSHub/pull/11487#discussion_r1055497421

https://github.com/DIYgod/RSSHub/blob/e66734aeca19ffa474f75b339b7f23fc570f1fb7/lib/v2/mrdx/daily.js#L61-L68

TonyRL commented on lib/v2/mrdx/daily.js:
```diff
@@ -0,0 +1,80 @@
+        passages.push({
+            title: passage.title,
+            author: passage.author,
+            pubDate: timezone(parseDate(passage.pubDate), 8),
```
Consider using RSSHub's utils `parseDate()` instead of reinventing the wheel **with wrong results**
docs: https://docs.rsshub.app/joinus/pub-date.html

![image](https://user-images.githubusercontent.com/11386903/209343974-88344097-c71a-42dc-86df-51b5783b9288.png)

Dustin-Jiang committed:
Merge branch 'DIYgod:master' into xinhua-daily-news

Dustin-Jiang committed:
feat(route): fix wrong date and improper cache method

Fix wrong `pubDate` by using `parseDate` in `utils/parse-date`.
Fix improper cache method. Now it caches proceeded RSS results.

TonyRL commented on lib/v2/mrdx/daily.js:
```diff
@@ -0,0 +1,79 @@
+                    title,
+                    author,
+                    pubDate: parseDate(pubDate, 'YYYY年MM月DD日'),
+                    guid: link,
+                    link,
```
> guid must be unique. link (defined from line 14) is not unique.

https://github.com/DIYgod/RSSHub/pull/11487#discussion_r1055499984
```suggestion
                    link: passage.link,
```

TonyRL replied:
This issue needs to be addressed before merging.
As you saw in the bot's preview, every feed item shares the **same** guid.
<details>
<summary>Feed 1</summary>

![1](https://user-images.githubusercontent.com/11386903/209466592-d537fae0-9502-4012-a9d7-bccb621b3d3a.PNG)
</details>
<details>
<summary>Feed 2</summary>

![2](https://user-images.githubusercontent.com/11386903/209466595-9ecb128a-2004-4eca-bdbf-78c011e4b69c.PNG)
</details>
<details>
<summary>Feed 3</summary>

![3](https://user-images.githubusercontent.com/11386903/209466597-69f6104f-65fa-4a5e-8337-18ba525edaf4.PNG)
</details>

Dustin-Jiang replied:
My bad. Now fixed.

TonyRL commented on lib/v2/mrdx/radar.js:
```diff
@@ -0,0 +1,13 @@
+            {
+                title: '今日',
+                docs: 'https://docs.rsshub.app/traditional-media.html#xin-hua-mei-ri-dian-xun',
+                source: '',
```
```suggestion
                source: ['*'],
```
You can add a wildcard match

Dustin-Jiang committed:
feat(radar): update source wildcard in lib/v2/mrdx/radar.js

Co-authored-by: Tony <TonyRL@users.noreply.github.com>

Dustin-Jiang committed:
fix(router): use unique article link as guid

Use specific article link as guid.
Use specific article link for every article instead of the homepage.
